internal/service/componentdescriptor/service.go

```diff
@@ -0,0 +1,223 @@
+				commonErrMsg(ocmi), err)
+	}
+
+	if ocmArtifactConfig.ComponentDescriptorLayer == nil {
+		return nil, fmt.Errorf("%w for %s", ErrLayerNil, commonErrMsg(ocmi))
+	}
+
+	compDescLayerDigest := ocmArtifactConfig.ComponentDescriptorLayer.Digest
+	if string(compDescLayerDigest) == "" {
+		return nil,
+			fmt.Errorf("%w for %s", ErrLayerDigestEmpty, commonErrMsg(ocmi))
+	}
+
+	layer, err := s.ociRepository.PullLayer(ctx, ocmi.Name(), ocmi.Version(), string(compDescLayerDigest))
+	if err != nil {
+		return nil, fmt.Errorf("failed to pull layer for ocm artifact with name=%q, version=%q and digest=%q: %w",
+			ocmi.Name(), ocmi.Version(), string(compDescLayerDigest), err)
+	}
+
+	ioh := &defaultExtractFileIOHelper{readAllFunc: io.ReadAll}
+	compdescBytes, err := extractFile(ioh, layer, ComponentDescriptorFileName)
+	if err != nil {
+		return nil,
+			fmt.Errorf("failed to extract component descriptor from layer fetched from %s with digest=%q: %w",
+				commonErrMsg(ocmi), string(compDescLayerDigest), err)
+	}
+
+	descriptor, err := deserialize(compdescBytes, ocmi)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.Descriptor{
+		ComponentDescriptor: descriptor,
+	}, nil
+}
+
+// deserialize decodes the component descriptor from its serialized form.
+func deserialize(compdescBytes []byte, ocmi ocmidentity.Component) (*compdesc.ComponentDescriptor, error) {
+	desc, err := compdesc.Decode(compdescBytes)
+	if err != nil {
+		return nil, fmt.Errorf("%w fetched from %s: %w",
+			ErrDecode, commonErrMsg(ocmi), err)
+	}
+	return desc, nil
+}
+
+// SimplifiedLayer represents only necessary part of containerregistryv1.Layer to simplify testing.
+type SimplifiedLayer interface {
+	Digest() (containerregistryv1.Hash, error)
+	Uncompressed() (io.ReadCloser, error)
+}
+
+func extractFile(ioh extractFileIOHelper, layer SimplifiedLayer, fileName string) ([]byte, error) {
+	wrap := func(err error) error {
+		digest, derr := layer.Digest()
+		if derr != nil {
+			err = errors.Join(err, derr)
+		}
+		return fmt.Errorf("failed to extract data of file=%q from TAR archive in a layer with digest=%q: %w",
+			fileName, digest, err)
+	}
+
+	layerReader, err := layer.Uncompressed()
+	if err != nil {
+		return nil, wrap(err)
+	}
+	defer layerReader.Close()
+
+	layerBytes, err := ioh.ReadAll(layerReader)
+	if err != nil {
+		return nil, wrap(err)
+	}
+
+	if len(layerBytes) == 0 {
+		return nil, wrap(ErrLayerEmpty)
+	}
+
+	tarIOh := &defaultUntarIOHelper{tarReader: tar.NewReader(bytes.NewReader(layerBytes))}
+	compdescBytes, err := unTar(tarIOh, fileName)
+	if err != nil {
+		return nil, wrap(err)
+	}
+
+	return compdescBytes, nil
+}
+
+type extractFileIOHelper interface {
+	ReadAll(r io.Reader) ([]byte, error) // Reads all data from the reader
+}
+
+type defaultExtractFileIOHelper struct {
+	readAllFunc func(r io.Reader) ([]byte, error)
+}
+
+func (d *defaultExtractFileIOHelper) ReadAll(r io.Reader) ([]byte, error) {
+	return d.readAllFunc(r)
+}
```

Comment:
same comment as for the wrapper in ocirepo.go. If we don't decide to go that we, I would at least name it consistently with Wrapper or Helper because eventually it is the same thing.

Reply:
They are gone. 👍